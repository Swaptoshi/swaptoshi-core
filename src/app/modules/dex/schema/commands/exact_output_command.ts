export const exactOutputCommandSchema = {
	$id: '/dex/command/exact_output',
	type: 'object',
	properties: {
		path: {
			dataType: 'bytes',
			format: 'hex',
			fieldNumber: 1,
		},
		recipient: {
			dataType: 'bytes',
			format: 'klayr32',
			fieldNumber: 2,
		},
		deadline: {
			dataType: 'string',
			fieldNumber: 3,
		},
		amountOut: {
			dataType: 'string',
			fieldNumber: 4,
		},
		amountInMaximum: {
			dataType: 'string',
			fieldNumber: 5,
		},
	},
};
