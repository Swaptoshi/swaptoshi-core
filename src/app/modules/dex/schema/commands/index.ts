export * from './burn_command';
export * from './collect_command';
export * from './collect_treasury_command';
export * from './create_pool_command';
export * from './decrease_liquidity_command';
export * from './exact_input_command';
export * from './exact_input_single_command';
export * from './exact_output_command';
export * from './exact_output_single_command';
export * from './increase_liquidity_command';
export * from './mint_command';
export * from './treasurify_command';
