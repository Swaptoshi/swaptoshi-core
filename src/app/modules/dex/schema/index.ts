/* eslint-disable import/no-cycle */
export * from './commands';
export * from './dependencies';
export * from './endpoint';
export * from './events';
export * from './nft_attribute';
export * from './stores';
export * from './config';
export * from './genesis';
export * from './nft';
