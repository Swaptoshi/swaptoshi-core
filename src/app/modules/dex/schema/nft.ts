const LENGTH_NFT_ID = 16;
const MIN_LENGTH_MODULE_NAME = 1;
const MAX_LENGTH_MODULE_NAME = 32;
const LENGTH_COLLECTION_ID = 4;

export const genesisNFTStoreSchema = {
	$id: '/nft/module/genesis',
	type: 'object',
	required: ['nftSubstore', 'supportedNFTsSubstore'],
	properties: {
		nftSubstore: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['nftID', 'owner', 'attributesArray'],
				properties: {
					nftID: {
						dataType: 'bytes',
						minLength: LENGTH_NFT_ID,
						maxLength: LENGTH_NFT_ID,
						fieldNumber: 1,
					},
					owner: {
						dataType: 'bytes',
						fieldNumber: 2,
					},
					attributesArray: {
						type: 'array',
						fieldNumber: 3,
						items: {
							type: 'object',
							required: ['module', 'attributes'],
							properties: {
								module: {
									dataType: 'string',
									minLength: MIN_LENGTH_MODULE_NAME,
									maxLength: MAX_LENGTH_MODULE_NAME,
									pattern: '^[a-zA-Z0-9]*$',
									fieldNumber: 1,
								},
								attributes: {
									dataType: 'bytes',
									fieldNumber: 2,
								},
							},
						},
					},
				},
			},
		},
		supportedNFTsSubstore: {
			type: 'array',
			fieldNumber: 2,
			items: {
				type: 'object',
				required: ['chainID', 'supportedCollectionIDArray'],
				properties: {
					chainID: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					supportedCollectionIDArray: {
						type: 'array',
						fieldNumber: 2,
						items: {
							type: 'object',
							required: ['collectionID'],
							properties: {
								collectionID: {
									dataType: 'bytes',
									minLength: LENGTH_COLLECTION_ID,
									maxLength: LENGTH_COLLECTION_ID,
									fieldNumber: 1,
								},
							},
						},
					},
				},
			},
		},
	},
};
