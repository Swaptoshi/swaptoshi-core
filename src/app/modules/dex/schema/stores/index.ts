export * from './observation';
export * from './pool';
export * from './position_info';
export * from './position_manager';
export * from './supported_token';
export * from './tick_bitmap';
export * from './tick_info';
export * from './token_symbol';
