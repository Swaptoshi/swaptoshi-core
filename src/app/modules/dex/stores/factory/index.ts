/* eslint-disable import/no-cycle */
import { NFTMethod, NamedRegistry, TokenMethod } from 'lisk-sdk';
import {
	SwaptoshiPoolData,
	DexModuleConfig,
	PositionManager,
	ImmutableSwapContext,
	MutableSwapContext,
} from '../../types';
import { SwaptoshiPool } from './pool';
import { SwapRouter } from './swap_router';
import { NonfungiblePositionManager } from './position_manager';

export function createImmutablePoolInstance(
	context: ImmutableSwapContext,
	pool: SwaptoshiPoolData,
	stores: NamedRegistry,
	events: NamedRegistry,
	tokenMethod: TokenMethod,
	config: DexModuleConfig,
) {
	const res = new SwaptoshiPool(pool, stores, events, config);
	res.addImmutableDependencies(context, tokenMethod);
	return res;
}

export function createMutablePoolInstance(
	context: MutableSwapContext,
	pool: SwaptoshiPoolData,
	stores: NamedRegistry,
	events: NamedRegistry,
	tokenMethod: TokenMethod,
	config: DexModuleConfig,
) {
	const res = new SwaptoshiPool(pool, stores, events, config);
	res.addMutableDependencies(context, tokenMethod);
	return res;
}

export function createMutableRouterInstance(
	context: MutableSwapContext,
	stores: NamedRegistry,
	tokenMethod: TokenMethod,
	config: DexModuleConfig,
) {
	const res = new SwapRouter(stores, config);
	res.addDependencies(context, tokenMethod);
	return res;
}

export function createImmutablePositionManagerinstance(
	positionManager: PositionManager,
	context: ImmutableSwapContext,
	stores: NamedRegistry,
	events: NamedRegistry,
	tokenMethod: TokenMethod,
	nftMethod: NFTMethod,
	chainId: Buffer,
) {
	const res = new NonfungiblePositionManager(positionManager, stores, events, chainId);
	res.addImmutableDependencies(context, tokenMethod, nftMethod);
	return res;
}

export function createMutablePositionManagerinstance(
	positionManager: PositionManager,
	context: MutableSwapContext,
	stores: NamedRegistry,
	events: NamedRegistry,
	tokenMethod: TokenMethod,
	nftMethod: NFTMethod,
	chainId: Buffer,
) {
	const res = new NonfungiblePositionManager(positionManager, stores, events, chainId);
	res.addMutableDependencies(context, tokenMethod, nftMethod);
	return res;
}

export { SwaptoshiPool } from './pool';
export { NonfungiblePositionManager } from './position_manager';
export { SwapRouter } from './swap_router';
