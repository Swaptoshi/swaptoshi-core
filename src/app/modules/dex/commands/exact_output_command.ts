/* eslint-disable class-methods-use-this */

import { BaseCommand, CommandVerifyContext, CommandExecuteContext, VerificationResult, VerifyStatus } from 'klayr-sdk';
import { exactOutputCommandSchema } from '../schema';
import { PoolStore } from '../stores/pool';
import { commandSwapContext } from '../stores/context';
import { ExactOutputParams } from '../types';
import { verifyExactOutputParam } from '../utils';

export class ExactOutputCommand extends BaseCommand {
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(_context: CommandVerifyContext<ExactOutputParams>): Promise<VerificationResult> {
		try {
			verifyExactOutputParam(_context.params);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async execute(_context: CommandExecuteContext<ExactOutputParams>): Promise<void> {
		const poolStore = this.stores.get(PoolStore);
		const context = commandSwapContext(_context);
		const router = poolStore.getMutableRouter(context);
		await router.exactOutput(_context.params);
	}

	public schema = exactOutputCommandSchema;
}
