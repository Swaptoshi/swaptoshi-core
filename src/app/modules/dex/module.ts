/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import {
	BaseInteroperableModule,
	BlockExecuteContext,
	FeeMethod,
	MainchainInteroperabilityMethod,
	ModuleInitArgs,
	ModuleMetadata,
	NFTMethod,
	SidechainInteroperabilityMethod,
	TokenMethod,
	TransactionExecuteContext,
	TransactionVerifyContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { BurnCommand } from './commands/burn_command';
import { CollectCommand } from './commands/collect_command';
import { CreatePoolCommand } from './commands/create_pool_command';
import { DecreaseLiquidityCommand } from './commands/decrease_liquidity_command';
import { ExactInputCommand } from './commands/exact_input_command';
import { ExactInputSingleCommand } from './commands/exact_input_single_command';
import { ExactOutputCommand } from './commands/exact_output_command';
import { ExactOutputSingleCommand } from './commands/exact_output_single_command';
import { IncreaseLiquidityCommand } from './commands/increase_liquidity_command';
import { MintCommand } from './commands/mint_command';
import { TreasurifyCommand } from './commands/treasurify_command';
import { DexEndpoint } from './endpoint';
import { BurnEvent } from './events/burn';
import { CollectEvent } from './events/collect';
import { CollectPositionEvent } from './events/collect_position';
import { CollectProtocolEvent } from './events/collect_protocol';
import { DecreaseLiquidityEvent } from './events/decrease_liquidity';
import { FlashEvent } from './events/flash';
import { IncreaseLiquidityEvent } from './events/increase_liquidity';
import { IncreaseObservationCardinalityNextEvent } from './events/increase_observation_cardinality_next';
import { MintEvent } from './events/mint';
import { PoolCreatedEvent } from './events/pool_created';
import { PoolInitializedEvent } from './events/pool_initialized';
import { SwapEvent } from './events/swap';
import { TokenURICreatedEvent } from './events/tokenuri_created';
import { TokenURIDestroyedEvent } from './events/tokenuri_destroyed';
import { DexMethod } from './method';
import {
	getPoolEndpointRequestSchema,
	getPoolEndpointResponseSchema,
} from './schema/endpoint/get_pool';
import {
	getPositionEndpointRequestSchema,
	getPositionEndpointResponseSchema,
} from './schema/endpoint/get_position';
import {
	getTokenURIEndpointRequestSchema,
	getTokenURIEndpointResponseSchema,
} from './schema/endpoint/get_token_uri';
import {
	observeEndpointRequestSchema,
	observeEndpointResponseSchema,
} from './schema/endpoint/observe';
import {
	quoteExactInputEndpointRequestSchema,
	quoteExactInputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input';
import {
	quoteExactInputSingleEndpointRequestSchema,
	quoteExactInputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input_single';
import {
	quoteExactOutputEndpointRequestSchema,
	quoteExactOutputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output';
import {
	quoteExactOutputSingleEndpointRequestSchema,
	quoteExactOutputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output_single';
import { ObservationStore } from './stores/observation';
import { PoolStore } from './stores/pool';
import { PositionInfoStore } from './stores/position_info';
import { PositionManagerStore } from './stores/position_manager';
import { TickBitmapStore } from './stores/tick_bitmap';
import { TickInfoStore } from './stores/tick_info';
import { TokenSymbolStore } from './stores/token_symbol';
import { DexModuleConfig } from './types';
import { verifyMinimumFee, verifySwapByTransfer, executeSwapByTransfer } from './hooks';
import { defaultConfig } from './constants';
import {
	getMetadataEndpointRequestSchema,
	getMetadataEndpointResponseSchema,
} from './schema/endpoint/get_metadata';
import { verifyFeeConversion } from './hooks/verifyFeeConversion';
import { executeFeeConversion } from './hooks/executeFeeConversion';
import { TreasurifyEvent } from './events/treasurify';
import {
	getPoolAddressFromCollectionIdEndpointRequestSchema,
	getPoolAddressFromCollectionIdEndpointResponseSchema,
} from './schema/endpoint/get_pool_address_from_collection_id';
import {
	quotePriceEndpointRequestSchema,
	quotePriceEndpointResponseSchema,
} from './schema/endpoint/quote_price';
import { TokenRegisteredEvent } from './events/token_registered';
import { SupportedTokenStore } from './stores/supported_token';
import { verifyValidTransfer } from './hooks/verifyValidTransfer';
import { DexInteroperableMethod } from './cc_method';

export class DexModule extends BaseInteroperableModule {
	public _config: DexModuleConfig | undefined;
	public _feeMethod: FeeMethod | undefined;
	public _tokenMethod: TokenMethod | undefined;
	public _dexInteroperableMethod = new DexInteroperableMethod(this.stores, this.events);

	public crossChainCommand = [];
	public crossChainMethod = this._dexInteroperableMethod;

	public endpoint = new DexEndpoint(this.stores, this.offchainStores);
	public method = new DexMethod(this.stores, this.events);
	public commands = [
		new CreatePoolCommand(this.stores, this.events),
		new MintCommand(this.stores, this.events),
		new IncreaseLiquidityCommand(this.stores, this.events),
		new DecreaseLiquidityCommand(this.stores, this.events),
		new CollectCommand(this.stores, this.events),
		new BurnCommand(this.stores, this.events),
		new ExactInputCommand(this.stores, this.events),
		new ExactInputSingleCommand(this.stores, this.events),
		new ExactOutputCommand(this.stores, this.events),
		new ExactOutputSingleCommand(this.stores, this.events),
		new TreasurifyCommand(this.stores, this.events),
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(PoolStore, new PoolStore(this.name, 0, this.stores, this.events));
		this.stores.register(
			PositionManagerStore,
			new PositionManagerStore(this.name, 1, this.stores, this.events),
		);

		this.stores.register(ObservationStore, new ObservationStore(this.name, 2));
		this.stores.register(PositionInfoStore, new PositionInfoStore(this.name, 3));
		this.stores.register(TickBitmapStore, new TickBitmapStore(this.name, 4));
		this.stores.register(TickInfoStore, new TickInfoStore(this.name, 5));
		this.stores.register(TokenSymbolStore, new TokenSymbolStore(this.name, 6, this.events));
		this.stores.register(SupportedTokenStore, new SupportedTokenStore(this.name, 7));

		this.events.register(BurnEvent, new BurnEvent(this.name));
		this.events.register(CollectPositionEvent, new CollectPositionEvent(this.name));
		this.events.register(CollectProtocolEvent, new CollectProtocolEvent(this.name));
		this.events.register(CollectEvent, new CollectEvent(this.name));
		this.events.register(DecreaseLiquidityEvent, new DecreaseLiquidityEvent(this.name));
		this.events.register(FlashEvent, new FlashEvent(this.name));
		this.events.register(IncreaseLiquidityEvent, new IncreaseLiquidityEvent(this.name));
		this.events.register(
			IncreaseObservationCardinalityNextEvent,
			new IncreaseObservationCardinalityNextEvent(this.name),
		);
		this.events.register(MintEvent, new MintEvent(this.name));
		this.events.register(PoolCreatedEvent, new PoolCreatedEvent(this.name));
		this.events.register(PoolInitializedEvent, new PoolInitializedEvent(this.name));
		this.events.register(SwapEvent, new SwapEvent(this.name));
		this.events.register(TokenURICreatedEvent, new TokenURICreatedEvent(this.name));
		this.events.register(TokenURIDestroyedEvent, new TokenURIDestroyedEvent(this.name));
		this.events.register(TreasurifyEvent, new TreasurifyEvent(this.name));
		this.events.register(TokenRegisteredEvent, new TokenRegisteredEvent(this.name));
	}

	public addDependencies(
		tokenMethod: TokenMethod,
		nftMethod: NFTMethod,
		feeMethod: FeeMethod,
		interoperabilityMethod: SidechainInteroperabilityMethod | MainchainInteroperabilityMethod,
	) {
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportManagerStore = this.stores.get(SupportedTokenStore);

		poolStore.addDependencies(tokenMethod);
		positionManagerStore.addDependencies(tokenMethod, nftMethod);
		supportManagerStore.addDependencies(tokenMethod);

		this._feeMethod = feeMethod;
		this._tokenMethod = tokenMethod;
		this._dexInteroperableMethod.addDependencies(interoperabilityMethod, tokenMethod, nftMethod);
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				{
					name: this.endpoint.getPool.name,
					request: getPoolEndpointRequestSchema,
					response: getPoolEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPosition.name,
					request: getPositionEndpointRequestSchema,
					response: getPositionEndpointResponseSchema,
				},
				{
					name: this.endpoint.getTokenURI.name,
					request: getTokenURIEndpointRequestSchema,
					response: getTokenURIEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPoolAddressFromCollectionId.name,
					request: getPoolAddressFromCollectionIdEndpointRequestSchema,
					response: getPoolAddressFromCollectionIdEndpointResponseSchema,
				},
				{
					name: this.endpoint.getMetadata.name,
					request: getMetadataEndpointRequestSchema,
					response: getMetadataEndpointResponseSchema,
				},
				{
					name: this.endpoint.observe.name,
					request: observeEndpointRequestSchema,
					response: observeEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInput.name,
					request: quoteExactInputEndpointRequestSchema,
					response: quoteExactInputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInputSingle.name,
					request: quoteExactInputSingleEndpointRequestSchema,
					response: quoteExactInputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutput.name,
					request: quoteExactOutputEndpointRequestSchema,
					response: quoteExactOutputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutputSingle.name,
					request: quoteExactOutputSingleEndpointRequestSchema,
					response: quoteExactOutputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quotePrice.name,
					request: quotePriceEndpointRequestSchema,
					response: quotePriceEndpointResponseSchema,
				},
			],
			assets: [],
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(_args: ModuleInitArgs): Promise<void> {
		this._config = { ...defaultConfig, ..._args.moduleConfig };
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportedTokenStore = this.stores.get(SupportedTokenStore);

		poolStore.init(this._config);
		positionManagerStore.init(Buffer.from(_args.genesisConfig.chainID, 'hex'));
		supportedTokenStore.init(this._config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		try {
			await verifyMinimumFee(_context, this._config!);
			await verifyValidTransfer.bind(this)(_context);
			await verifySwapByTransfer.bind(this)(_context);
			await verifyFeeConversion.bind(this)(_context);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await verifyMinimumFee(_context, this._config!);
		await verifyValidTransfer.bind(this)(_context);
		await verifySwapByTransfer.bind(this)(_context);
		await executeFeeConversion.bind(this)(_context);
	}

	public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await executeSwapByTransfer.bind(this)(_context);
	}

	public async beforeTransactionsExecute(context: BlockExecuteContext): Promise<void> {
		await this.stores.get(SupportedTokenStore).apply(context);
	}
}
