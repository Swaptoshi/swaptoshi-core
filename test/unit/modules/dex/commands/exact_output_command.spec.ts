/* eslint-disable camelcase */
/* eslint-disable jest/expect-expect */
import { StateMachine } from 'klayr-sdk';
import { DexModule } from '../../../../../src/app/modules/dex/module';
import { ExactOutputParams, TokenMethod } from '../../../../../src/app/modules/dex/types';
import { invalidAddress, invalidNumberString } from '../utils/invalid';
import { commandFixture } from '../utils/fixtures';
import { NonfungiblePositionManager } from '../../../../../src/app/modules/dex/stores/factory';
import { FeeAmount, TICK_SPACINGS, getMaxTick, getMinTick } from '../stores/shared/utilities';
import { poolAddress, senderAddress, senderPublicKey, token0, token1 } from '../utils/account';
import { NFTRegistry } from '../stores/shared/nft/nft_registry';
import { TokenRegistry } from '../stores/shared/token/token_registry';
import { exactOutputCommandSchema } from '../../../../../src/app/modules/dex/schema';
import { ExactOutputCommand } from '../../../../../src/app/modules/dex/commands/exact_output_command';
import { encodePath } from '../stores/shared/path';
import { Uint } from '../../../../../src/app/modules/dex/stores/library/int';
import { mock_token_transfer } from '../stores/shared/token';

type CommandParam = ExactOutputParams;
const COMMAND_NAME = 'exactOutput';
const commandSchema = exactOutputCommandSchema;

const validParam: CommandParam = {
	path: encodePath([token1, token0], new Array<string>(1).fill(FeeAmount.MEDIUM)),
	recipient: senderAddress,
	deadline: Date.now().toString(),
	amountOut: '1',
	amountInMaximum: '3',
};

describe('ExactOutputCommand', () => {
	let module: DexModule;
	let command: ExactOutputCommand;
	let nft: NonfungiblePositionManager;
	let tokenMethod: TokenMethod;
	let createCommandVerifyContext: (params: CommandParam) => StateMachine.CommandVerifyContext<CommandParam>;
	let createCommandExecuteContext: (params: CommandParam) => StateMachine.CommandExecuteContext<CommandParam>;

	beforeEach(async () => {
		({ module, createCommandExecuteContext, createCommandVerifyContext, nft, tokenMethod } = await commandFixture<CommandParam>(COMMAND_NAME, commandSchema, senderPublicKey, validParam));
		command = new ExactOutputCommand(module.stores, module.events);

		await nft.mint({
			token0,
			token1,
			fee: FeeAmount.MEDIUM,
			tickLower: getMinTick(TICK_SPACINGS[FeeAmount.MEDIUM]).toString(),
			tickUpper: getMaxTick(TICK_SPACINGS[FeeAmount.MEDIUM]).toString(),
			recipient: senderAddress,
			amount0Desired: '1000000',
			amount1Desired: '1000000',
			amount0Min: '0',
			amount1Min: '0',
			deadline: Date.now().toString(),
		});
	});

	afterEach(() => {
		NFTRegistry.reset();
		TokenRegistry.reset();
	});

	const getBalances = async (who: Buffer, context: StateMachine.CommandExecuteContext<CommandParam>) => {
		const balances = await Promise.all([await tokenMethod.getAvailableBalance(context, who, token0), await tokenMethod.getAvailableBalance(context, who, token1)]);
		return {
			token0: Uint.from(balances[0]),
			token1: Uint.from(balances[1]),
		};
	};

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toBe(COMMAND_NAME);
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		it('should return status OK when called with valid input', async () => {
			const context = createCommandVerifyContext(validParam);
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.OK);
		});

		it('should throw error when user sends transaction with invalid address (recipient)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				recipient: invalidAddress,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (deadline)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				deadline: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (amountOut)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				amountOut: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (amountInMaximum)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				amountInMaximum: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});
	});

	describe('execute', () => {
		it('should swap tokens successfully', async () => {
			const context = createCommandExecuteContext(validParam);

			const poolBefore = await getBalances(poolAddress, context);
			const senderBefore = await getBalances(senderAddress, context);

			await command.execute(context);

			const poolAfter = await getBalances(poolAddress, context);
			const senderAfter = await getBalances(senderAddress, context);

			expect(senderAfter.token0).toStrictEqual(senderBefore.token0.sub(3));
			expect(senderAfter.token1).toStrictEqual(senderBefore.token1.add(1));
			expect(poolAfter.token0).toStrictEqual(poolBefore.token0.add(3));
			expect(poolAfter.token1).toStrictEqual(poolBefore.token1.sub(1));
		});

		it('should add command events', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			expect(mock_token_transfer).toHaveBeenCalledWith(senderAddress.toString('hex'), poolAddress.toString('hex'), '3');

			expect(mock_token_transfer).toHaveBeenCalledWith(poolAddress.toString('hex'), senderAddress.toString('hex'), '1');
		});
	});
});
