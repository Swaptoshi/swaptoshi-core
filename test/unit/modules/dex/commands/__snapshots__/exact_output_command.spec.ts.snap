// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExactOutputCommand constructor should have valid schema 1`] = `
{
  "$id": "/dex/command/exact_output",
  "properties": {
    "amountInMaximum": {
      "dataType": "string",
      "fieldNumber": 5,
    },
    "amountOut": {
      "dataType": "string",
      "fieldNumber": 4,
    },
    "deadline": {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "path": {
      "dataType": "bytes",
      "fieldNumber": 1,
      "format": "hex",
    },
    "recipient": {
      "dataType": "bytes",
      "fieldNumber": 2,
      "format": "klayr32",
    },
  },
  "type": "object",
}
`;
